app/controllers/vm_cloud_controller.rb

```diff
@@ -309,6 +307,87 @@ def cancel_action(message)
     replace_right_cell
   end
 
+  def evacuate
+    assert_privileges("instance_evacuate")
+    @record = find_by_id_filtered(VmOrTemplate, params[:id]) # Set the VM object
+    if @record.is_available?(:evacuate) && !@record.ext_management_system.nil?
+      drop_breadcrumb(
+        :name => _("Evacuate Instance '%{name}'") % {:name => @record.name},
+        :url  => "/vm_cloud/evacuate"
+      ) unless @explorer
+      @in_a_form = true
+      @refresh_partial = "vm_common/evacuate"
+    else
+      add_flash(_("Unable to evacuate %{instance} \"%{name}\": %{details}") % {
+        :instance => ui_lookup(:table => 'vm_cloud'),
+        :name     => @record.name,
+        :details  => @record.is_available_now_error_message(:evacuate)}, :error)
+    end
+  end
+  alias_method :instance_evacuate, :evacuate
+
+  def evacuate_form_fields
+    assert_privileges("instance_evacuate")
+    @record = find_by_id_filtered(VmOrTemplate, params[:id])
+    host_cluster = @record.try!(:host).try!(:ems_cluster)
+    hosts = []
+    if host_cluster
+      hosts = host_cluster.hosts.reject { |h| h == @record.host }.map do |h|
+        {:id => h.id, :name => h.name}
+      end
+    end
+    render :json => {
+      :hosts => hosts
+    }
+  end
+
+  def evacuate_vm
+    assert_privileges("instance_evacuate")
+    @record = find_by_id_filtered(VmOrTemplate, params[:id])
+
+    case params[:button]
+    when "cancel"
+      cancel_action(_("Evacuation of %{model} \"%{name}\" was cancelled by the user") % {
+        :model => ui_lookup(:table => 'vm_cloud'),
+        :name  => @record.name
+      })
+    when "submit"
+      if @record.is_available?(:evacuate)
+        if params['auto_select_host'] == 'on'
+          hostname = nil
+        else
+          hostname = find_by_id_filtered(Host, params[:destination_host_id]).hostname
```

Comment:
We need to make sure :destination_host_id is not nil. If it's nil, the user needs an error message in the form saying the field is required. More generally, any failure in form submission needs to display to the user -- we should avoid ever returning the form to the user on a failure without a message of some sort.